src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -160,67 +163,120 @@ private BoundExpression VisitIndexerAccess(BoundIndexerAccess node, bool isLeftO
             {
-                return VisitIndexPatternIndexerAccess(
-                    node.Syntax,
-                    node.Receiver,
-                    node.LengthOrCountProperty,
-                    (PropertySymbol)node.PatternSymbol,
-                    node.Argument,
-                    isLeftOfAssignment: isLeftOfAssignment);
+                return VisitIndexPatternIndexerAccess(node, isLeftOfAssignment: isLeftOfAssignment);
             }
             else
             {
                 Debug.Assert(TypeSymbol.Equals(
                     node.Argument.Type,
                     _compilation.GetWellKnownType(WellKnownType.System_Range),
                     TypeCompareKind.ConsiderEverything));
-                return VisitRangePatternIndexerAccess(
-                    node.Receiver,
-                    node.LengthOrCountProperty,
-                    (MethodSymbol)node.PatternSymbol,
-                    node.Argument);
+                Debug.Assert(!isLeftOfAssignment || node.IndexerAccess.GetRefKind() == RefKind.Ref);
+
+                return VisitRangePatternIndexerAccess(node);
             }
         }
 
-
-        private BoundSequence VisitIndexPatternIndexerAccess(
-            SyntaxNode syntax,
-            BoundExpression receiver,
-            PropertySymbol lengthOrCountProperty,
-            PropertySymbol intIndexer,
-            BoundExpression argument,
-            bool isLeftOfAssignment)
+        private BoundExpression VisitIndexPatternIndexerAccess(BoundIndexOrRangePatternIndexerAccess node, bool isLeftOfAssignment)
         {
+            Debug.Assert(node.ArgumentPlaceholders.Length == 1);
+            Debug.Assert(node.IndexerAccess is BoundIndexerAccess);
+
+            Debug.Assert(TypeSymbol.Equals(
+                node.Argument.Type,
+                _compilation.GetWellKnownType(WellKnownType.System_Index),
+                TypeCompareKind.ConsiderEverything));
+
             var F = _factory;
 
             var locals = ArrayBuilder<LocalSymbol>.GetInstance(2);
             var sideeffects = ArrayBuilder<BoundExpression>.GetInstance(2);
 
-            Debug.Assert(receiver.Type is { });
-            var receiverLocal = F.StoreToTemp(
-                VisitExpression(receiver),
-                out var receiverStore,
-                // Store the receiver as a ref local if it's a value type to ensure side effects are propagated
-                receiver.Type.IsReferenceType ? RefKind.None : RefKind.Ref);
-            locals.Add(receiverLocal.LocalSymbol);
-            sideeffects.Add(receiverStore);
+            var receiver = VisitExpression(node.GetReceiver());
+
+            // Do not capture receiver if it is a local or parameter and we are evaluating a pattern
+            // If length access is a local, then we are evaluating a pattern
+            if (node.LengthOrCountAccess.Kind is not BoundKind.Local || receiver.Kind is not BoundKind.Local or BoundKind.Parameter)
```

Comment:
@AlekseyTs Would a more general handling like `CanChangeValueBetweenReads` work here?

Reply:
> Would a more general handling like CanChangeValueBetweenReads work here? 

I don't think we can rely on that because in general a local can change between reads. Assigned in an index value, for example. 

> Maybe we would encode the knowledge that temps from patterns cannot be captured or change between reads.

I am not sure it is worth coming up with a more general solution because we only need to handle patters and we know what to expect from them here.